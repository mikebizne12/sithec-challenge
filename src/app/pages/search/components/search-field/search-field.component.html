<form [formGroup]="form" (ngSubmit)="onSearch()" class="grid">
  <div class="col-12 md:col-12 md:col-3">
    <span class="p-float-label">
      <input pInputText id="name" formControlName="name" class="w-full" />
      <label htmlFor="name">Name</label>
    </span>
  </div>
  <div
    class="col-12 md:col-3"
    [hidden]="filter == LOCATION || filter == EPISODE"
  >
    <p-dropdown
      [options]="status"
      optionLabel="name"
      optionValue="id"
      [style]="{ width: '100%' }"
      formControlName="status"
      placeholder="Status"
    />
  </div>
  <div
    class="col-12 md:col-3"
    [hidden]="filter == LOCATION || filter == EPISODE"
  >
    <span class="p-float-label">
      <input pInputText id="species" formControlName="species" class="w-full" />
      <label htmlFor="species">Species</label>
    </span>
  </div>

  <div class="col-12 md:col-3" [hidden]="filter == EPISODE">
    <span class="p-float-label">
      <input pInputText id="type" formControlName="type" class="w-full" />
      <label htmlFor="type">Type</label>
    </span>
  </div>
  <div
    class="col-12 md:col-3"
    [hidden]="filter == CHARACTER || filter == EPISODE"
  >
    <span class="p-float-label">
      <input
        pInputText
        id="dimension"
        formControlName="dimension"
        class="w-full"
      />
      <label htmlFor="dimension">Dimension</label>
    </span>
  </div>
  <div
    class="col-12 md:col-3"
    [hidden]="filter == CHARACTER || filter == LOCATION"
  >
    <span class="p-float-label">
      <input pInputText id="episode" formControlName="episode" class="w-full" />
      <label htmlFor="episode">Episode</label>
    </span>
  </div>
  <div
    class="col-12 md:col-3"
    [hidden]="filter == LOCATION || filter == EPISODE"
  >
    <p-dropdown
      [options]="gender"
      optionLabel="name"
      [style]="{ width: '100%' }"
      optionValue="id"
      formControlName="gender"
      placeholder="Gender"
    />
  </div>
  <div class="col-6 md:col-3">
    <p-button
      type="submit"
      icon="pi pi-search"
      label="Search"
      [style]="{ width: '100%' }"
      [disabled]="form.invalid"
    ></p-button>
  </div>
  <div class="col-6 md:col-3">
    <p-button
      type="submit"
      icon="pi pi-eraser"
      severity="secondary"
      label="Clear"
      [style]="{ width: '100%' }"
      (onClick)="resetForm()"
    ></p-button>
  </div>
</form>
