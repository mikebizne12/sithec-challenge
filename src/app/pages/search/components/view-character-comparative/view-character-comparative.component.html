<div class="grid">
  <div class="text-center col-12" *ngIf="loading">
    <p-progressSpinner></p-progressSpinner>
  </div>
  <div class="col" *ngFor="let item of data; let i = index">
    <p-card
      header="{{ item.name | truncate : 17 }}"
      [style]="{ width: 'auto' }"
    >
      <ng-template pTemplate="header">
        <img
          alt="Card"
          [defaultImage]="'assets/imgs/no-image.png'"
          [lazyLoad]="item.image"
        />
      </ng-template>

      <p-chip label="{{ item.status }} - {{ item.species }}"></p-chip>
      <p><strong>Gender: </strong>{{ item.gender }}</p>
      <p><b>Location: </b>{{ item.location!.name }}</p>
      <p><b>Species: </b>{{ item.species }}</p>
      <p><b>Status: </b>{{ item.status }}</p>
      <p><b>Episodes: </b>{{ item.episode.length }}</p>
      <p><b>Shared episodes:</b></p>
      <ul>
        <ng-container *ngFor="let char of charactersCommom">
          <li *ngIf="char.id != item.id">
            <p>
              {{ char.name }} -
              {{ getEpisodesCommon(data[i].episode, char.episodes).length }}
              episode/s
            </p>
          </li>
        </ng-container>
      </ul>
    </p-card>
  </div>
</div>
