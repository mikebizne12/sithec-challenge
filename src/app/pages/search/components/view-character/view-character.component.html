|<p-button
  type="submit"
  label="Compare selected"
  icon="pi pi-arrow-right-arrow-left"
  id="button-compare"
  (onClick)="viewCompared()"
  [disabled]="getIds().length == 1 || getIds().length === 0"
></p-button>
<p-toast position="bottom-center"></p-toast>
<div class="grid fadein animation-duration-1000">
  <div *ngFor="let item of data.results" class="col-12 md:col-4 lg:col-3">
    <p-card header="{{ item.name | truncate: 17 }}" [style]="{ width: 'auto' }">
      <ng-template pTemplate="header">
        <img
          alt="Card"
          [defaultImage]="'assets/imgs/no-image.png'"
          [lazyLoad]="item.image"
        />
      </ng-template>

      <p-chip label="{{ item.status }} - {{ item.species }}"></p-chip>
      <p>
        <i class="pi pi-globe"></i>
        {{ item.location!.name | truncate: 20 }}
      </p>
      <p *ngIf="item.episode_detail">
        <i class="pi pi-bookmark"></i>
        {{ item.episode_detail.episode }} |
        {{ item.episode_detail.air_date | date }}
      </p>
      <ng-template pTemplate="footer">
        <p-button
          label="View"
          icon="pi pi-eye"
          (onClick)="selectCharacter(item)"
        ></p-button>
        <p-button
          label="Compare"
          [icon]="'pi ' + (item.compared ? 'pi-check' : '')"
          class="ml-2"
          (onClick)="validCheck(item)"
          iconPos="right"
          styleClass="p-button-secondary"
          [disabled]="item.disabled || false"
        ></p-button>
      </ng-template>
    </p-card>
  </div>
</div>
