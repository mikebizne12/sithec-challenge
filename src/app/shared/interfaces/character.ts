import { IEpisode } from './episode';
import { ILocation } from './location';

export interface ICharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type?: string;
  gender?: string;
  origin: any;
  location: ILocation;
  image: string;
  episode: Array<string>;
  episode_detail: IEpisode;
}
