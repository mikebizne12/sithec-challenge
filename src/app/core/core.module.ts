import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { HttpTokenInterceptor } from './interceptors/http.token.interceptor';
import { ErrorHandlerInterceptor } from './interceptors/error-handler.interceptor';
import { ErrorHandlersUtilitiesService } from './utilities/error-handlers-utilities.service';
import { ToastModule } from 'primeng/toast';

@NgModule({
  imports: [CommonModule, RouterModule, ToastModule],
  declarations: [],
  exports: [],
  providers: [
    ErrorHandlersUtilitiesService,
    { provide: HTTP_INTERCEPTORS, useClass: HttpTokenInterceptor, multi: true },
    {
      provide: HTTP_INTERCEPTORS,
      useClass: ErrorHandlerInterceptor,
      multi: true,
    },
  ],
})
export class CoreModule {}
