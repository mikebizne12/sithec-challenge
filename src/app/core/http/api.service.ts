import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';
import { IConfig } from '../../shared/interfaces/config';
import { Observable } from 'rxjs';

@Injectable()
export class ApiService {
  private url = '';

  constructor(public http: HttpClient) {
    this.url = `${environment.api}`;
  }

  get(endpoint: string, params?: any, reqOpts?: any) {
    if (!reqOpts) {
      reqOpts = {
        params: new HttpParams(),
      };
    }
    if (params) {
      reqOpts.params = new HttpParams();
      for (const k in params) {
        reqOpts.params = reqOpts.params.set(k, params[k]);
      }
    }
    return this.http.get<IConfig>(this.url + '/' + endpoint, reqOpts);
  }
}
