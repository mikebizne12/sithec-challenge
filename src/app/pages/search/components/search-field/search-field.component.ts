import { Component, EventEmitter, Input, Output } from '@angular/core';
import { UntypedFormGroup } from '@angular/forms';
import { IList } from 'src/app/shared/interfaces/list';

@Component({
  selector: 'app-search-field',
  templateUrl: './search-field.component.html',
  styleUrls: ['./search-field.component.scss'],
})
export class SearchFieldComponent {
  @Input() form!: UntypedFormGroup;
  @Input() filter = 1;
  @Output() eventClic: EventEmitter<any> = new EventEmitter();

  gender: Array<IList> = [
    { id: 'female', name: 'Female' },
    { id: 'male', name: 'Male' },
    { id: 'genderless', name: 'Genderless' },
    { id: 'unknown', name: 'Unknown' },
  ];

  status: Array<IList> = [
    { id: 'alive', name: 'Alive' },
    { id: 'dead ', name: 'Dead ' },
    { id: 'unknown', name: 'Unknown' },
  ];

  CHARACTER = 1;
  LOCATION = 2;
  EPISODE = 3;

  onSearch() {
    this.eventClic.emit();
  }

  resetForm() {
    this.form.reset();
    this.form.controls['page'].setValue(1);
    this.form.controls['selector'].setValue(1);
  }
}
