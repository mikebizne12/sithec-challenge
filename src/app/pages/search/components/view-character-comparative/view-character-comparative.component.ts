import { Component, OnInit } from '@angular/core';
import { ICharacter } from 'src/app/shared/interfaces/character';
import { CharacterService } from '../../../../core/services/character.service';
import { DynamicDialogConfig } from 'primeng/dynamicdialog';

@Component({
  selector: 'app-view-character-comparative',
  templateUrl: './view-character-comparative.component.html',
  styleUrls: ['./view-character-comparative.component.scss'],
})
export class ViewCharacterComparativeComponent implements OnInit {
  ids: Array<number> = [];
  loading = true;
  data: Array<ICharacter> = [];
  charactersCommom: any = [];
  constructor(
    private characterService: CharacterService,
    private dialogConfig: DynamicDialogConfig,
  ) {}

  ngOnInit(): void {
    this.ids = this.dialogConfig.data.ids;

    this.characterService
      .getMultiplerCharacter(this.ids)
      .subscribe((resp: any) => {
        resp.forEach((element: any) => {
          this.charactersCommom.push({
            id: element.id,
            name: element.name,
            episodes: element.episode,
          });
        });
        this.data = resp;
        this.loading = false;
      });
  }

  getEpisodesCommon(basedata: Array<string>, comparedata: Array<string>) {
    const common: any = [];
    basedata.filter((d: any) => {
      comparedata.filter((s: any) => {
        if (d === s) {
          common.push(d);
        }
      });
    });
    return common;
  }
}
