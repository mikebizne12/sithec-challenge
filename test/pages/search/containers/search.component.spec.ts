import { ComponentFixture, TestBed } from '@angular/core/testing';
import { SearchComponent } from '../../../../src/app/pages/search/containers/search/search.component';
import { CharacterService } from '../../../../src/app/core/services/character.service';
import { ApiService } from '../../../../src/app/core/http/api.service';
import { HttpClientModule } from '@angular/common/http';
import { ViewCharacterComponent } from '../../../../src/app/pages/search/components/view-character/view-character.component';
import { ViewLocationComponent } from '../../../../src/app/pages/search/components/view-location/view-location.component';
import { ViewEpisodeComponent } from '../../../../src/app/pages/search/components/view-episode/view-episode.component';
import { ViewEpisodeCharacterComponent } from '../../../../src/app/pages/search/components/view-episode-character/view-episode-character.component';
import { ViewCharacterComparativeComponent } from '../../../../src/app/pages/search/components/view-character-comparative/view-character-comparative.component';
import { SearchFieldComponent } from '../../../../src/app/pages/search/components/search-field/search-field.component';

import {
  FormBuilder,
  Validators,
  ReactiveFormsModule,
  FormsModule,
} from '@angular/forms';

describe('SearchComponent', () => {
  let component: SearchComponent;
  let fixture: ComponentFixture<SearchComponent>;
  const formBuilder: FormBuilder = new FormBuilder();
  let compiled: HTMLElement;
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [HttpClientModule, ReactiveFormsModule, FormsModule],
      declarations: [
        SearchComponent,
        SearchFieldComponent,
        ViewCharacterComponent,
        ViewLocationComponent,
        ViewEpisodeComponent,
        ViewEpisodeCharacterComponent,
        ViewCharacterComparativeComponent,
      ],
      providers: [
        CharacterService,
        ApiService,
        { provide: FormBuilder, useValue: formBuilder },
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(SearchComponent);
    component = fixture.componentInstance;
    component.formSearch = formBuilder.group({
      selector: [1, Validators.required],
      name: [''],
      status: [''],
      species: [''],
      type: [''],
      gender: [''],
      dimension: [''],
      episode: [''],
      page: [1],
    });
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should match with snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
