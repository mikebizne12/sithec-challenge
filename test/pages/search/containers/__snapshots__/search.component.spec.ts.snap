// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent should match with snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="container"
  >
    <div
      class="text-center"
    >
      <img
        class="w-50"
        src="/assets/imgs/title.png"
      />
    </div>
    <div
      class="row pt-4"
    >
      <div
        class="col-12 col-md-2 p-0 mb-3"
      >
        <div
          aria-label="Button group with nested dropdown"
          class="btn-group dropdown"
          ngbdropdown=""
          role="group"
        >
          <button
            aria-expanded="false"
            class="dropdown-toggle btn btn-primary"
            ngbdropdowntoggle=""
          >
             Character 
            <fa-icon
              class="ng-fa-icon ml-2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 416c-8.188 0-16.38-3.125-22.62-9.375l-192-192c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L224 338.8l169.4-169.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-192 192C240.4 412.9 232.2 416 224 416z"
                  fill="currentColor"
                />
              </svg>
            </fa-icon>
          </button>
          <div
            class="dropdown-menu"
            ngbdropdownmenu=""
            x-placement="bottom"
          >
            <button
              class="dropdown-item"
              ngbdropdownitem=""
            >
               Character 
            </button>
            <button
              class="dropdown-item"
              ngbdropdownitem=""
            >
               Location 
            </button>
            <button
              class="dropdown-item"
              ngbdropdownitem=""
            >
               Episode 
            </button>
            
          </div>
        </div>
      </div>
      <div
        class="col-12 col-md-10 p-0"
      >
        <app-search-field>
          <form
            class="row px-2 mb-4 ng-untouched ng-pristine ng-valid"
            novalidate=""
          >
            <div
              class="form-group col-12 col-md-3"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="name"
                id="name"
                placeholder="Name"
                type="text"
              />
            </div>
            <div
              class="form-group col-12 col-md-3"
            >
              <select
                class="form-control rounded-0 ng-untouched ng-pristine ng-valid"
                formcontrolname="status"
              >
                <option
                  ngvalue=""
                  value="0: "
                >
                  Any status
                </option>
                <option
                  value="1: alive"
                >
                   Alive 
                </option>
                <option
                  value="2: dead "
                >
                   Dead  
                </option>
                <option
                  value="3: unknown"
                >
                   Unknown 
                </option>
                
              </select>
            </div>
            <div
              class="form-group col-12 col-md-3"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="species"
                id="species"
                placeholder="Species"
                type="text"
              />
            </div>
            <div
              class="form-group col-12 col-md-3"
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="type"
                id="type"
                placeholder="Type"
                type="text"
              />
            </div>
            <div
              class="form-group col-12 col-md-3"
              hidden=""
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="dimension"
                id="dimension"
                placeholder="Dimension"
                type="text"
              />
            </div>
            <div
              class="form-group col-12 col-md-3"
              hidden=""
            >
              <input
                class="form-control ng-untouched ng-pristine ng-valid"
                formcontrolname="episode"
                id="episode"
                placeholder="Episode"
                type="text"
              />
            </div>
            <div
              class="form-group col-12 col-md-3"
            >
              <select
                class="form-control rounded-0 ng-untouched ng-pristine ng-valid"
                formcontrolname="gender"
              >
                <option
                  ngvalue=""
                  value="0: "
                >
                  Any gender
                </option>
                <option
                  value="1: female"
                >
                   Female 
                </option>
                <option
                  value="2: male"
                >
                   Male 
                </option>
                <option
                  value="3: genderless"
                >
                   Genderless 
                </option>
                <option
                  value="4: unknown"
                >
                   Unknown 
                </option>
                
              </select>
            </div>
            <div
              class="form-group"
            >
              <button
                class="btn btn-primary"
                type="submit"
              >
                <fa-icon
                  class="ng-fa-icon"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-magnifying-glass"
                    data-icon="magnifying-glass"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M500.3 443.7l-119.7-119.7c27.22-40.41 40.65-90.9 33.46-144.7C401.8 87.79 326.8 13.32 235.2 1.723C99.01-15.51-15.51 99.01 1.724 235.2c11.6 91.64 86.08 166.7 177.6 178.9c53.8 7.189 104.3-6.236 144.7-33.46l119.7 119.7c15.62 15.62 40.95 15.62 56.57 0C515.9 484.7 515.9 459.3 500.3 443.7zM79.1 208c0-70.58 57.42-128 128-128s128 57.42 128 128c0 70.58-57.42 128-128 128S79.1 278.6 79.1 208z"
                      fill="currentColor"
                    />
                  </svg>
                </fa-icon>
              </button>
            </div>
          </form>
        </app-search-field>
      </div>
    </div>
    
    <app-view-character>
      <button
        class="btn btn-primary mb-4"
        disabled=""
        type="button"
      >
         Compare selected

      </button>
      <div
        class="row"
      >
        
      </div>
    </app-view-character>
    
    
    
    <div
      class="row m-auto"
    >
      <div
        class="col-12"
      >
        
      </div>
    </div>
  </div>
</div>
`;
